### PR TITLE
Add prefilter flag

### LINKED ISSUE #2539: Backspace generated by IME gets lost
Please fill in the following items if you don't know the root cause.

Which distribution and version?: Ubuntu 23.04

Which desktop environment and version?: GNOME 44.0

Which session type?: X11

Which application and version?:
All applications that don't support surrounding text,
e.g. Chrome/Chromium, gnome-terminal

IBus version?: 1.5.28

Issue description:
Backspace generated by the IME gets often lost

Steps to reproduce:
1. Install Chromium:
   ```bash
   sudo apt update
   sudo apt install chromium-browser
   ```

2. Install `keyman` package which will also install `ibus-keyman` and
   `python3-keyman-config` packages:
   ```bash
   sudo apt install keyman
   sudo reboot
   ```

3. Install the "IPA (SIL)" Keyman keyboard by running
   ```bash
   km-package-install -p sil_ipa
   ```
   This will add a "Undetermined (IPA (SIL))"
   keyboard to the keyboard dropdown list.
4. Open https://www.editpad.org/ in Chromium
5. switch to "Undetermined (IPA (SIL))" keyboard
6. In a new text note, type `n` followed by `>`

**Expected output:**
`ŋ`

**Actual output:**
`nŋ`

(When the user presses `>` ibus-engine-keyman sends a backspace followed by `ŋ`.
This should delete `n` and replace it with `ŋ`. However, the backspace
seems to get lost. See verbose output below.)

Can you reproduce your problem when you restart ibus-daemon? (yes / no):
yes

Do you see any errors when you run ibus-daemon with the verbose option?: No. See output below.

Can you reproduce your problem with a new user account instead of the current your account? (yes / no): yes (after keyboard installation for that user)

<details>
   <summary>Output of running `ibus-daemon --xim --verbose`</summary>

```
** Message: 17:50:23.795: adding kmp path /home/user/.local/share/keyman/sil_ipa
** Message: 17:50:23.795: could not find website in json

** Message: 17:50:23.938: adding engine und-Latn:/home/user/.local/share/keyman/sil_ipa/sil_ipa.kmx
** Message: 17:50:44.693: adding kmp path /home/user/.local/share/keyman/sil_ipa
** Message: 17:50:44.693: could not find website in json

** Message: 17:50:44.836: adding engine und-Latn:/home/user/.local/share/keyman/sil_ipa/sil_ipa.kmx
** Message: 17:50:44.837: WDG: km_service_init
** Message: 17:50:44.842: DAR: ibus_keyman_engine_constructor und-Latn:/home/user/.local/share/keyman/sil_ipa/sil_ipa.kmx
** Message: 17:50:44.842: ibus_keyman_engine_constructor: Loading options for kb_name: sil_ipa
** Message: 17:50:44.842: keyman_get_options_queue_fromdconf
** Message: 17:50:44.842: keyman_get_options_fromdconf

(ibus-engine-keyman:4390): GLib-CRITICAL **: 17:50:44.843: g_queue_foreach: assertion 'func != NULL' failed
** Message: 17:50:44.843: reset_context
** Message: 17:50:44.845: ibus_keyman_engine_focus_in
** Message: 17:50:44.845: reset_context
** Message: 17:50:44.845: WDG: ibus_keyman_engine_enable und-Latn:/home/user/.local/share/keyman/sil_ipa/sil_ipa.kmx: enabling surrounding context
** Message: 17:50:46.344: -----------------------------------------------------------------------------------------------------------------
** Message: 17:50:46.344: DAR: ibus_keyman_engine_process_key_event - keyval=0x6e keycode=0x31, state=0x10, isKeyDown=1, supports_prefilter=0
** Message: 17:50:46.344: ibus_keyman_engine_process_key_event: before process key event
** Message: 17:50:46.344: get_current_context_text: current context is:0:1::
** Message: 17:50:46.344: DAR: ibus_keyman_engine_process_key_event - km_mod_state=0x0
** Message: 17:50:46.345: ibus_keyman_engine_process_key_event: after process key event
** Message: 17:50:46.345: get_current_context_text: current context is:1:2:n:
** Message: 17:50:46.345: CHAR action 1/2
** Message: 17:50:46.345: process_unicode_char_action: unichar:U+006e, bytes:1, string:n
** Message: 17:50:46.345: process_unicode_char_action: setting buffer to converted unichar
** Message: 17:50:46.345: process_unicode_char_action: CHAR buffer is now n
** Message: 17:50:46.345: END action 2/2
** Message: 17:50:46.345: DAR: commit_string - n
** Message: 17:50:46.345: ibus_keyman_engine_process_key_event: after processing all actions
** Message: 17:50:46.345: get_current_context_text: current context is:1:2:n:
** Message: 17:50:46.419: -----------------------------------------------------------------------------------------------------------------
** Message: 17:50:46.419: DAR: ibus_keyman_engine_process_key_event - keyval=0x6e keycode=0x31, state=0x40000010, isKeyDown=0, supports_prefilter=0
** Message: 17:50:46.419: ibus_keyman_engine_process_key_event: before process key event
** Message: 17:50:46.419: get_current_context_text: current context is:1:2:n:
** Message: 17:50:46.419: DAR: ibus_keyman_engine_process_key_event - km_mod_state=0x0
** Message: 17:50:46.420: ibus_keyman_engine_process_key_event: after process key event
** Message: 17:50:46.420: get_current_context_text: current context is:1:2:n:
** Message: 17:50:46.420: EMIT_KEYSTROKE action 1/2
** Message: 17:50:46.420: END action 2/2
** Message: 17:50:47.642: -----------------------------------------------------------------------------------------------------------------
** Message: 17:50:47.642: DAR: ibus_keyman_engine_process_key_event - keyval=0xffe1 keycode=0x2a, state=0x10, isKeyDown=1, supports_prefilter=0
** Message: 17:50:47.642: ibus_keyman_engine_process_key_event: before process key event
** Message: 17:50:47.642: get_current_context_text: current context is:1:2:n:
** Message: 17:50:47.642: DAR: ibus_keyman_engine_process_key_event - km_mod_state=0x0
** Message: 17:50:47.643: ibus_keyman_engine_process_key_event: after process key event
** Message: 17:50:47.643: get_current_context_text: current context is:1:2:n:
** Message: 17:50:47.643: END action 1/1
** Message: 17:50:47.643: ibus_keyman_engine_process_key_event: after processing all actions
** Message: 17:50:47.643: get_current_context_text: current context is:1:2:n:
** Message: 17:50:48.552: -----------------------------------------------------------------------------------------------------------------
** Message: 17:50:48.552: DAR: ibus_keyman_engine_process_key_event - keyval=0x3e keycode=0x34, state=0x11, isKeyDown=1, supports_prefilter=0
** Message: 17:50:48.553: ibus_keyman_engine_process_key_event: before process key event
** Message: 17:50:48.553: get_current_context_text: current context is:1:2:n:
** Message: 17:50:48.553: DAR: ibus_keyman_engine_process_key_event - km_mod_state=0x10
** Message: 17:50:48.553: ibus_keyman_engine_process_key_event: after process key event
** Message: 17:50:48.553: get_current_context_text: current context is:1:3:ŋ:
** Message: 17:50:48.553: BACK action 1/3
** Message: 17:50:48.553: DAR: process_backspace_action - client_capabilities=29, 20
** Message: 17:50:48.553: process_backspace_action: deleting surrounding text 1 char
** Message: 17:50:48.553: CHAR action 2/3
** Message: 17:50:48.553: process_unicode_char_action: unichar:U+014b, bytes:2, string:ŋ
** Message: 17:50:48.553: process_unicode_char_action: setting buffer to converted unichar
** Message: 17:50:48.553: process_unicode_char_action: CHAR buffer is now ŋ
** Message: 17:50:48.553: END action 3/3
** Message: 17:50:48.553: DAR: commit_string - ŋ
** Message: 17:50:48.553: ibus_keyman_engine_process_key_event: after processing all actions
** Message: 17:50:48.553: get_current_context_text: current context is:1:3:ŋ:
** Message: 17:50:48.650: -----------------------------------------------------------------------------------------------------------------
** Message: 17:50:48.650: DAR: ibus_keyman_engine_process_key_event - keyval=0x3e keycode=0x34, state=0x40000011, isKeyDown=0, supports_prefilter=0
** Message: 17:50:48.650: ibus_keyman_engine_process_key_event: before process key event
** Message: 17:50:48.650: get_current_context_text: current context is:1:3:ŋ:
** Message: 17:50:48.650: DAR: ibus_keyman_engine_process_key_event - km_mod_state=0x10
** Message: 17:50:48.650: ibus_keyman_engine_process_key_event: after process key event
** Message: 17:50:48.650: get_current_context_text: current context is:1:3:ŋ:
** Message: 17:50:48.650: EMIT_KEYSTROKE action 1/2
** Message: 17:50:48.650: END action 2/2
** Message: 17:50:49.896: -----------------------------------------------------------------------------------------------------------------
** Message: 17:50:49.896: DAR: ibus_keyman_engine_process_key_event - keyval=0xffe1 keycode=0x2a, state=0x40000011, isKeyDown=0, supports_prefilter=0
** Message: 17:50:49.896: ibus_keyman_engine_process_key_event: before process key event
** Message: 17:50:49.896: get_current_context_text: current context is:1:3:ŋ:
** Message: 17:50:49.896: DAR: ibus_keyman_engine_process_key_event - km_mod_state=0x10
** Message: 17:50:49.896: ibus_keyman_engine_process_key_event: after process key event
** Message: 17:50:49.896: get_current_context_text: current context is:1:3:ŋ:
** Message: 17:50:49.896: END action 1/1
** Message: 17:50:49.896: ibus_keyman_engine_process_key_event: after processing all actions
** Message: 17:50:49.896: get_current_context_text: current context is:1:3:ŋ:
** Message: 17:50:52.146: ibus_keyman_engine_reset
** Message: 17:50:52.146: reset_context
** Message: 17:50:52.146: reset_context: new context is :: (len:0) cursor:0 anchor:0
** Message: 17:50:52.146: get_current_context_text: current context is:1:3:ŋ:
** Message: 17:50:52.146: reset_context: setting context because it has changed from expected
** Message: 17:50:52.147: ibus_keyman_engine_focus_out
** Message: 17:50:52.148: ibus_keyman_engine_focus_in
** Message: 17:50:52.148: reset_context
** Message: 17:50:52.148: reset_context: new context is :: (len:0) cursor:0 anchor:0
** Message: 17:50:52.149: get_current_context_text: current context is:0:1::
** Message: 17:50:52.149: reset_context: setting context because it has changed from expected
** Message: 17:50:52.152: ibus_keyman_engine_focus_out
** Message: 17:50:52.152: ibus_keyman_engine_focus_in
** Message: 17:50:52.152: reset_context
** Message: 17:50:52.152: reset_context: new context is :: (len:0) cursor:0 anchor:0
** Message: 17:50:52.152: get_current_context_text: current context is:0:1::
** Message: 17:50:52.152: reset_context: setting context because it has changed from expected
** Message: 17:50:52.152: ibus_keyman_engine_focus_out
** Message: 17:50:52.153: ibus_keyman_engine_focus_in
** Message: 17:50:52.153: reset_context
** Message: 17:50:52.153: reset_context: new context is :: (len:0) cursor:0 anchor:0
** Message: 17:50:52.153: get_current_context_text: current context is:0:1::
** Message: 17:50:52.153: reset_context: setting context because it has changed from expected
** Message: 17:50:52.828: ibus_keyman_engine_focus_out
** Message: 17:50:52.828: WDG: ibus_keyman_engine_disable und-Latn:/home/user/.local/share/keyman/sil_ipa/sil_ipa.kmx
** Message: 17:50:52.828: ibus_keyman_engine_focus_out
** Message: 17:50:52.828: DAR: ibus_keyman_engine_destroy und-Latn:/home/user/.local/share/keyman/sil_ipa/sil_ipa.kmx
```
</details>

### References

Other manifestations of this bug:

- https://github.com/keymanapp/keyman/issues/1489
- https://github.com/keymanapp/keyman/issues/4028
- https://github.com/keymanapp/keyman/issues/4029
- https://github.com/keymanapp/keyman/issues/4030
- https://github.com/keymanapp/keyman/issues/4505

Note: the issues are closed because they would be fixed by #2440.

### PR DESCRIPTION
**Pull request by ermshiperete**
This PR adds a new `IBUS_PREFILTER_MASK` flag which the engine can set on the state. This allows engines to prefilter events and use that to control the order of the output even if some syntethic key events get interspersed between the input.

A scenario where this is necessary is in [ibus-engine-keyman](https://github.com/keymanapp/keyman) for applications that don't support surrounding text like Chrome/Chromium and gnome-terminal. For example with the [_IPA (SIL)_](https://keyman.com/keyboards/sil_ipa) keyboard, the users presses the `n` key which produces 'n' output. Then the user presses `;`. The expected result is that the 'n' gets replaced by 'ŋ'. For applications that don't support surrounding text, the engine has to forward a Backspace key event followed by committing the new text 'ŋ'. However, since the key event gets processed asynchronously, it can happen that the new text gets committed before the Backspace key is processed, leading to the deletion of 'ŋ' instead of 'n'.

The `IBUS_PREFILTER_MASK` flag allows to solve this problem: when the engine receives a key press, it calls `ibus_engine_forward_key_event()` for the same key with the `IBUS_PREFILTER_MASK` flag set. Only when it receives a key press with the `IBUS_PREFILTER_MASK` already set it will process the key. When it generates the Backspace key event it will forward it to ibus with the `IBUS_PREFILTER_MASK` flag already set. When it comes back to the engine the key events will be in the correct order.

Since `IBUS_PREFILTER_MASK` is a flag that will only ever get set by the engine it shouldn't break any existing engines.

This PR also adds `IBUS_CAP_PREFILTER` to `IBusCapabilite`.  This allows engines to detect whether they are dealing with an ibus version that was compiled with prefilter support.

Fixes #2539.

### CODE REVIEW

**Comment by ermshiperete** [comment]
@fujiwarat Would it be possible to get this PR merged? We really need it to get the above mentioned scenario working. Thanks!

**Comment by fujiwarat** [comment]
Now is the bug fix phase for the stable repository and I'm a bit busy. After Fedora 37 is released, I will bring back to this issue.
Thank you.

**Comment by ermshiperete** [comment]
@fujiwarat Now with Fedora 37 released, could this PR get merged?

File: client/gtk2/ibusimcontext.c [hunk 1]
```diff
@@ -682,8 +683,10 @@ _key_snooper_cb (GtkWidget   *widget,
     if (G_UNLIKELY (event->state & IBUS_HANDLED_MASK))
         return TRUE;
 
-    if (G_UNLIKELY (event->state & IBUS_IGNORED_MASK))
-        return FALSE;
+    if (!(event->state & IBUS_PREFILTER_MASK)) {
```
Comment (fujiwarat) [inline client/gtk2/ibusimcontext.c:686]: I disagree to handle the return value with the configurable flag.

File: client/gtk2/ibusimcontext.c [hunk 2]
```diff
@@ -1179,7 +1182,7 @@ ibus_im_context_filter_keypress (GtkIMContext *context,
         GdkModifierType state = gdk_event_get_modifier_state (event);
         if (state & IBUS_HANDLED_MASK)
             return TRUE;
-        if (state & IBUS_IGNORED_MASK)
+        if (state & IBUS_IGNORED_MASK && !(state & IBUS_PREFILTER_MASK))
```
Comment (fujiwarat) [inline client/gtk2/ibusimcontext.c:1190]: I disagree to handle the return value with the configurable flag.

**Review by fujiwarat — CHANGES_REQUESTED** [review]
I don't think this kind of the logic.

**Review by ermshiperete — COMMENTED** [review]
Thanks for your review! I'm not sure I completely understand your comments.

Let me try to explain again:

For events that don't have the `IBUS_PREFILTER_MASK` set we want the processing to be the same as before. Which means basically if the `IBUS_HANDLED_MASK` or `IBUS_IGNORED_MASK` bits are set, the event gets send to the client. If those bits are not set the event gets send to the engine. (This might be oversimplistic, but shows the general direction).

If the `IBUS_PREFILTER_MASK` is set, we want the event to always end up at the engine. It's the engine's responsibility to set the flag if desired, and to also clear the flag when it's no longer needed. It's also the engine's responsibility to check the capabilities and only set the `IBUS_PREFILTER_MASK` if the `IBUS_CAP_PREFILTER` capability is set.

It might be counter-intuitive that we always check that `IBUS_PREFILTER_MASK` is not set. The code would be easier to understand if we could check for the presence of a flag instead of the absence. Unfortunately that's not possible if we want to keep backwards-compatibility, because it would mean that all engines would have to set the new flag.

It might be surprising that we don't check `IBUS_CAP_PREFILTER` in the ibus code itself (except for setting the capabilities). That's an optimization that I try to explain below.

**Comment by fujiwarat** [comment]
> A scenario where this is necessary is in [ibus-engine-keyman](https://github.com/keymanapp/keyman) for applications that don't support surrounding text like Chrome/Chromium and gnome-terminal. For example with the [IPA (SIL)](https://keyman.com/keyboards/sil_ipa) keyboard, the users presses the n key which produces 'n' output. Then the user presses ;. The expected result is that the 'n' gets replaced by 'ŋ'. For applications that don't support surrounding tex

I guess your engine needs to be updated or your issue can be resolved in the ibus clients.
It would be good to report your issue with detail instead of creating a patch.

File: client/gtk2/ibusimcontext.c (hunk 2 above)
Comment (ermshiperete) [reply client/gtk2/ibusimcontext.c:1190]: Which implementation exactly?

**Comment by ermshiperete** [comment]
> > A scenario where this is necessary is in [ibus-engine-keyman](https://github.com/keymanapp/keyman) for applications that don't support surrounding text like Chrome/Chromium and gnome-terminal. For example with the [IPA (SIL)](https://keyman.com/keyboards/sil_ipa) keyboard, the users presses the n key which produces 'n' output. Then the user presses ;. The expected result is that the 'n' gets replaced by 'ŋ'. For applications that don't support surrounding tex
> 
> I guess your engine needs to be updated or your issue can be resolved in the ibus clients. It would be good to report your issue with detail instead of creating a patch.

**I would be very happy to report an issue instead of creating a patch.** That would save me a lot of time. But I doubt it would be fixed in the foreseeable future.

This can't be completely implemented in the engine nor in the client since it requires the coordination between input and output events, something only ibus can do.

**Comment by fujiwarat** [comment]
As I noted you should report your issue with detail at first instead pushing the wrong patch.

**Comment by ermshiperete** [comment]
> As I noted you should report your issue with detail at first instead pushing the wrong patch.

Sorry, only after I pressed the re-review button I noticed that I had neglected to create the requested issue. Did that now: #2539.

**Comment by fujiwarat** [comment]
Closing the wrong report which does not fix anything.
